lightningd/plugin.c

```diff
@@ -373,6 +407,182 @@ static bool plugin_opts_add(const struct plugin_request *req)
+
+	res = req->resulttok;
+	response = json_stream_success(rpc_req->cmd);
+
+	/* If this is a raw string result we'd be cropping the quotes from the
+	 * result, so add them back */
+	if (req->resulttok->type == JSMN_STRING)
+		json_add_member(response, NULL, "\"%.*s\"",
+				res->end - res->start,
+				req->response + res->start);
+	else
+		json_add_member(response, NULL, "%.*s", res->end - res->start,
+				req->response + res->start);
+
+	command_success(rpc_req->cmd, response);
+	tal_free(rpc_req);
+}
+
+static void plugin_rpcmethod_dispatch(struct command *cmd, const char *buffer,
+				      const jsmntok_t *params)
+{
+	const jsmntok_t *toks = params, *methtok, *idtok;
+	struct plugin_rpc_request *request;
+	struct plugins *plugins = cmd->ld->plugins;
+	struct plugin *plugin;
+
+	if (cmd->mode == CMD_USAGE) {
+		cmd->usage = "[params]";
+		return;
+	}
+
+	/* We're given the params, but we need to walk back to the
+	 * root object, so just walk backwards until the current
+	 * element has no parents, that's going to be the root
+	 * element. */
+	while (toks->parent != -1)
+		toks--;
+
+	methtok = json_get_member(buffer, toks, "method");
+	idtok = json_get_member(buffer, toks, "id");
+	/* We've parsed them before, these should not fail! */
+	assert(idtok != NULL && methtok != NULL);
+
+	request = tal(NULL, struct plugin_rpc_request);
+	request->method = tal_strndup(request, buffer + methtok->start,
+				      methtok->end - methtok->start);
+	request->id = tal_strndup(request, buffer + idtok->start,
+				      idtok->end - idtok->start);
+	request->params = tal_strndup(request, buffer + params->start,
+				      params->end - params->start);
+	request->plugin = NULL;
+	request->cmd = cmd;
+
+	/* Find the plugin that registered this RPC call */
+	list_for_each(&plugins->plugins, plugin, list) {
+		for (size_t i=0; i<tal_count(plugin->methods); i++) {
+			if (streq(request->method, plugin->methods[i])) {
+				request->plugin = plugin;
+				break;
+			}
+		}
+	}
+
+	/* This should never happen, it'd mean that a plugin didn't
+	 * cleanup after dying */
+	assert(request->plugin);
+
+	tal_steal(request->plugin, request);
+	plugin_request_send(request->plugin, request->method, request->params, plugin_rpcmethod_cb, request);
+
+	command_still_pending(cmd);
+}
+
+static bool plugin_rpcmethod_add(struct plugin *plugin, const char *buffer,
+				 const jsmntok_t *meth)
+{
+	const jsmntok_t *nametok, *desctok, *longdesctok;
+	struct json_command *cmd;
+
+	nametok = json_get_member(buffer, meth, "name");
+	desctok = json_get_member(buffer, meth, "description");
+	longdesctok = json_get_member(buffer, meth, "long_description");
+
+	if (!nametok || nametok->type != JSMN_STRING) {
+		plugin_kill(
+		    plugin,
+		    tal_fmt(plugin,
+			    "rpcmethod does not have a string \"name\": %.*s",
+			    meth->end - meth->start, buffer + meth->start));
+		return false;
+	}
+
+	if (!desctok || desctok->type != JSMN_STRING) {
+		plugin_kill(plugin, tal_fmt(plugin,
```

Comment:
meta comment that we should make sure we make clear in the docs that a description is mandatory